src/services/jsTyping.ts

```diff
@@ -0,0 +1,293 @@
+                (options.include && options.include.length > 0) ||
+                (options.exclude && options.exclude.length > 0)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * @param host is the object providing I/O related operations.
+     * @param fileNames are the file names that belong to the same project.
+     * @param globalCachePath is used to get the safe list file path and as cache path if the project root path isn't specified.
+     * @param projectRootPath is the path to the project root directory. This is used for the local typings cache.
+     * @param typingOptions are used for customizing the typing inference process.
+     * @param compilerOptions are used as a source of typing inference.
+     */
+    export function discoverTypings(
+        host: TypingResolutionHost,
+        fileNames: string[],
+        globalCachePath: Path,
+        projectRootPath: Path,
+        typingOptions: TypingOptions,
+        compilerOptions: CompilerOptions):
+        { cachedTypingPaths: string[], newTypingNames: string[], filesToWatch: string[] } {
+
+        // A typing name to typing file path mapping
+        const inferredTypings: Map<string> = {};
+
+        if (!isTypingEnabled(typingOptions)) {
+            return { cachedTypingPaths: [], newTypingNames: [], filesToWatch: [] };
+        }
+
+        const cachePath = projectRootPath || globalCachePath;
+        // Only infer typings for .js and .jsx files
+        fileNames = filter(map(fileNames, ts.normalizePath), f => scriptKindIs(f, /*LanguageServiceHost*/ undefined, ScriptKind.JS, ScriptKind.JSX));
+
+        const safeListFilePath = ts.combinePaths(globalCachePath, "safeList.json");
+        if (!safeList && host.fileExists(safeListFilePath)) {
+            safeList = tryParseJson(safeListFilePath, host);
+        }
+
+        const filesToWatch: string[] = [];
+        // Directories to search for package.json, bower.json and other typing information
+        let searchDirs: string[] = [];
+        let exclude: string[] = [];
+
+        mergeTypings(typingOptions.include);
+        exclude = typingOptions.exclude || [];
+
+        if (typingOptions.enableAutoDiscovery) {
+            const possibleSearchDirs = map(fileNames, ts.getDirectoryPath);
+            if (projectRootPath !== undefined) {
+                possibleSearchDirs.push(projectRootPath);
+            }
+            searchDirs = ts.deduplicate(possibleSearchDirs);
+            for (const searchDir of searchDirs) {
+                const packageJsonPath = ts.combinePaths(searchDir, "package.json");
+                getTypingNamesFromJson(packageJsonPath, filesToWatch);
+
+                const bowerJsonPath = ts.combinePaths(searchDir, "bower.json");
+                getTypingNamesFromJson(bowerJsonPath, filesToWatch);
+
+                const nodeModulesPath = ts.combinePaths(searchDir, "node_modules");

```

Comment:
so why do we do look at node_modules though we have found a package.json? but we do not look for bower_components?